Visible whitespace only for selections

tomKPZ opened the pull request.

tomKPZ commented:
This is my first foray into rust so the code may not be idomatic

the-mikedavis commented on helix-core/src/syntax.rs:
```diff
@@ -1830,6 +1832,7 @@ fn injection_for_match<'a>(
 pub struct Merge<I> {
     iter: I,
     spans: Box<dyn Iterator<Item = (usize, std::ops::Range<usize>)>>,
+    emit_selection_events: bool,
```
I'm not sure this would actually be necessary for future merges/iterators but I think it would be more elegant to have this be

```rs
pub struct Merge<I> {
    // ..
    events: Option<(HighlightEvent, HighlightEvent)>,
    //                  ^ start event     ^ end event
    // ..
}
```

passing `None` for merges like diagnostics where we don't need to emit events and `Some((SelectionStart, SelectionEnd))` for selections when selection whitespace rendering is enabled.

The boolean just for selection events seems very specific for this otherwise very generic merge function/type.

tomKPZ replied:
Thanks for your review.  I've rebased and implemented your suggestion in the latest revision.

the-mikedavis reviewed (commented):
Sorry for the delay on this. Admittedly it took me a while to figure out how this was working despite being familiar with these parts of the codebase 😅

the-mikedavis commented:
I'm not sure this makes sense or if it's too much of a micro optimization but we could not emit selection events when none of the WhitespaceRenderValues are for select. What do you think?

tomKPZ commented:
> I'm not sure this makes sense or if it's too much of a micro optimization but we could not emit selection events when none of the WhitespaceRenderValues are for select. What do you think?

Sure, done in latest revision

archseer commented on helix-core/src/syntax.rs:
```diff
@@ -1144,6 +1144,8 @@ pub enum HighlightEvent {
     Source { start: usize, end: usize },
     HighlightStart(Highlight),
     HighlightEnd,
+    SelectionStart,
+    SelectionEnd,
```
Sorry it took a while! I wonder if we could avoid additional event types and instead just track this in the render loop. When `HighlightStart(selection scope)` is emitted we'd toggle a local bool to true, then false when it's popped off the scope stack

tomKPZ replied:
Done.  I hope I've understood your suggestion properly

the-mikedavis replied:
I think archseer is saying that you can track this without changing the `helix_core::syntax` by looking up the scope for the highlight event and seeing if it's a selection scope:

```rs
HighlightEvent::HighlightStart(span) => {
    if theme
        .scopes()
        .get(span.0)
        .map(|scope| scope.starts_with("ui.selection"))
        .unwrap_default()
    {
        selecting = true;
    }
    spans.push(span);
}
// and for HighlightEnd, check that the popped scope starts with "ui.selection"
```

that snippet ☝️ isn't very performant though - we should calculate which indices in `theme.scopes()` correspond to selection highlights outside of the `for`

tomKPZ replied:
I see, thanks for clarifying.  I think I've done it in the latest revision

the-mikedavis commented:
It looks like whitespace ends up not getting rendered for the cursor. For that we may need to include `ui.cursor` and `ui.cursor.*` (but there's also `ui.cursorline` so we can't use a starts-with check)

archseer commented:
I've done some simplifications, this way we can avoid comparing spans against the set of selection scopes in most cases. I'm also thinking if we want to remove that render closure for a regular function

tomKPZ commented:
Thanks for the commits @archseer !

I've added a fix for `ui.cursor` that @the-mikedavis found, and I've removed the inner `render_whitespace` closure, so no more dynamic dispatch is required for each whitespace char rendered.

pascalkuthe commented:
With #5420 merged this PR will need to be ported to the new text rendering in `document.rs`

pascalkuthe commented:
clsoing this one as stale as the rendering code has been entriely replaced
